src/http/input/body/N3PatchBodyParser.ts

```diff
@@ -0,0 +1,136 @@
+  public async canHandle({ metadata }: BodyParserArgs): Promise<void> {
+    if (metadata.contentType !== TEXT_N3) {
+      throw new UnsupportedMediaTypeHttpError('This parser only supports N3 Patch data.');
+    }
+  }
+
+  public async handle({ request, metadata }: BodyParserArgs): Promise<N3Patch> {
+    const n3 = await readableToString(request);
+    const parser = new Parser({ format: TEXT_N3, baseIRI: metadata.identifier.value });
+    let store: Store;
+    try {
+      store = new Store(parser.parse(n3));
+    } catch (error: unknown) {
+      throw new BadRequestHttpError(`Invalid N3: ${createErrorMessage(error)}`);
+    }
+    const def = DataFactory.defaultGraph();
+
+    // Solid, §5.3.1: "A patch resource MUST contain a triple ?patch rdf:type solid:InsertDeletePatch."
+    // "The patch document MUST contain exactly one patch resource,
+    // identified by one or more of the triple patterns described above, which all share the same ?patch subject."
+    const subjects = store.getSubjects(RDF.terms.type, SOLID.terms.InsertDeletePatch, def);
+    if (subjects.length !== 1) {
+      throw new UnprocessableEntityHttpError('An N3 Patch document requires exactly 1 solid:InsertDeletePatch entry.');
+    }
+    const subject = subjects[0];
+
+    // Solid, §5.3.1: "A patch resource MUST be identified by a URI or blank node, which we refer to as ?patch
+    // in the remainder of this section."
+    if (subject.termType !== 'NamedNode' && subject.termType !== 'BlankNode') {
+      throw new UnprocessableEntityHttpError('An N3 Patch subject needs to be a blank or named node.');
+    }
+
+    // Extract all quads from the corresponding formulae
+    const deletes = this.findQuads(store, subject, SOLID.terms.deletes, def);
+    const inserts = this.findQuads(store, subject, SOLID.terms.inserts, def);
+    const conditions = this.findQuads(store, subject, SOLID.terms.where, def);
+
+    // Make sure there are no forbidden combinations
+    const conditionVars = this.findVariables(conditions);
+    this.verifyQuads(deletes, conditionVars, 'delete');
+    this.verifyQuads(inserts, conditionVars, 'insert');
+
+    return {
+      deletes,
+      inserts,
+      conditions,
+      binary: true,
+      data: guardedStreamFrom(n3),
+      metadata,
+      isEmpty: false,
+    };
+  }
+
+  /**
+   * Finds all quads in a where/deletes/inserts formula.
+   * The returned quads will be updated so their graph is the default graph instead of the N3 reference to the formula.
+   * Will error in case there are multiple instances of the subject/predicate combination.
+   */
+  private findQuads(store: Store, subject: Quad_Subject, predicate: NamedNode, def: DefaultGraph): Quad[] {
+    const graphs = store.getObjects(subject, predicate, def);
+    if (graphs.length > 1) {
+      throw new UnprocessableEntityHttpError(`An N3 Patch document can have at most 1 ${predicate.value}.`);
+    }
+    if (graphs.length === 0) {
+      return [];
+    }
```

Comment:
```suggestion
```

Reply:
`graphs.length` being 0 means there was no entry for that specific predicate (which corresponds to an empty array of triples). If not there would be an error on the next line as well trying to access `graphs[0]`.